Add utility script to merge Cachi2 and Syft SBOMs

brunoapimentel opened the pull request:
The script performs a deduplication of reported components, favoring Cachi2's report whenever possible:
- Removes any Syft component that has a matching PURL with a Cachi2 component
- Removes Golang locally replaced modules from Syft
- Removes non-registry Pip and NPM dependencies reported by Syft

More details about the deduplication can be found [here](https://github.com/brunoapimentel/rhtap-cachi2-sbom-merge/blob/main/find-duplicates.ipynb).

This script is primarily meant to perform the merging of both SBOMs in the RHTAP build pipeline. The reason to implement it in Cachi2 is because of its growing complexity, which would be rather unmaintanable if written inline in a Tekton yaml file.

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [n/a] Docs updated (if applicable)
- [n/a] Docs links in the code are still valid (if docs were updated)

chmeliik commented:
> Removes Pip non-Pypi dependencies reported by Syft

What about npm non-registry dependencies?

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+    Local replacements are reported in a very different way by Cachi2,
+    which is why the same reports by Syft should be removed.
+    """
+    return (
+        component.get("purl", "").startswith("pkg:golang")
+        and component.get("name", "").startswith(".")
+        or component.get("version", "") == "(devel)"
+    )
```
Depending on the intended behavior, this may need more parentheses
```python
In [1]: False and False or True
Out[1]: True
```

brunoapimentel replied:
Fixed.

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+    return (
+        component.get("purl", "").startswith("pkg:pypi")
+        and "version" not in component
+    )
```
Maybe checking for vcs_url | download_url qualifiers would be more robust and re-usable for npm?

brunoapimentel replied:
Good point. I did not do any research on how NPM dedup would look like, though. I'll look into this tomorrow, and I bet it add more dedup rules to this already long script :confused: 

We can revisit this topic then.

brunoapimentel replied:
Ok, so NPM rule is added.

Troblem about defaulting to `vcs_url | download_url` is that we only check for Cachi2 non-registry components for Pip. For NPM, we check for Syft non-registry components, and for Golang, we only care about local replacements.

brunoapimentel replied:
Updating this thread here: we have decided to use the same rule to dedup non-registry components for both NPM and Pip (that is, check Cachi2 reported non-registry components, and remove all Syft components that have a matching name). So, yeah, it makes sense to use `vcs_url | download_url` qualifiers.

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+    """Create a unique key from Cachi2 reported components."""
+    url = urlsplit(component["purl"])
+    # omit the query part, since Cachi2 sets it and Syft does not
+    return url.scheme + ":" + url.path
```
This also omits the fragment (subpath), that's probably intentional too?

brunoapimentel replied:
It is intentional, since Syft does not set any qualifiers or subpaths (for Golang and Pip, at least)

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+        parts = component["purl"].split("@")
+        # We need to encode special characters in the version part,
+        # and also need to ensure the name is all lowercase
+        return parts[0].lower() + "@" + quote_plus(parts[1])
```
Just FYI: not all [purl types](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) should lowercase the name. But those that cachi2 currently supports should.

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+
+    if "@" in component["purl"]:
+        parts = component["purl"].split("@")
+        # We need to encode special characters in the version part,
+        # and also need to ensure the name is all lowercase
+        return parts[0].lower() + "@" + quote_plus(parts[1])
```
> We need to encode special characters in the version part,

Here we take the `version?qualifiers#subpath` part. Encoding that as a whole breaks the purl, but I guess Syft never uses any qualifiers or subpath?

chmeliik replied:
Hmm, we already make that assumption [here](https://github.com/containerbuildsystem/cachi2/pull/223#discussion_r1259607142) so this is consistent

brunoapimentel replied:
I added more info on the docstrings to try to clear this up.

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+    This function applies several rules to determine if a component should be
+    removed from the final merged SBOM.
+    """
+    cachi2_non_pypi_components = [
+        component["name"] for component in cachi_sbom_components
+        if _is_cachi2_non_pypi_pip_dependency(component)
+    ]
```
Is it safe to de-duplicate by `name` only? While it's not possible to have two different versions of the same dependency in one requirements file, it's possible to have them across different req files or across two different packages

brunoapimentel replied:
Yeah, I thought about that. Since Cachi2 does not report the version for vcs/file dependencies, there's no way to accurately determine that it is exactly the same dependency. 

This won't be a problem if all packages are processed by Cachi2, since we'll get Cachi2's report of all versions of this dependency. The corner case I see is if one of the packages is not processed by Cachi2, or if it's a package lying somewhere in the image and Syft picks it up.

I think the alternative option here is to live with duplicates of Pip non-PiPY dependencies =/
Unless you have another idea on how to handle this.

chmeliik replied:
What Syft SBOMs will we be merging with this exactly? The one generated from the source directory? From the built image? Both?

brunoapimentel replied:
Both Syft-source and Syft-image SBOMs are already merged in the build pipeline nowadays. So we'll do another merge, and add Cachi2's SBOM there.

I added a Pip few examples [here](https://github.com/brunoapimentel/rhtap-cachi2-sbom-merge), but I still need to update the investigation as I did with Golang.

chmeliik replied:
Hmm, theoretically any dependencies unknown to cachi2 should only be able to get there from the base image. I'd say it's fine to drop the Syft duplicates by name in this case

brunoapimentel replied:
Ack. I will make our approach clearer in the docstrings.

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+    ]
+
+    syft_sbom["components"] = filtered_syft_components + cachi2_sbom["components"]
+    syft_sbom["metadata"]["tools"].append({
+        "vendor": "Red Hat",
+        "name": "Cachi2"
+    })
```
IMO this doesn't belong in a merge script but in cachi2 itself

brunoapimentel replied:
Agreed. We will need to do some merging magic here regardless.

brunoapimentel replied:
I changed the script so it will merge any `metadata.tools` if they're available. The second part should be solved by #232.

chmeliik commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        print("Usage: ./merge_syft_sbom.py path/to/cachi2.sbom path/to/syft.sbom")
```
should be len < 3

maybe also exit with non-zero

or use argparse :)

brunoapimentel replied:
I keep forgeting Python has these types of magic. Changed to argparse.

brunoapimentel commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
```
Added.

chmeliik commented on tests/unit/data/sboms/merged.bom.json:
```diff
@@ -0,0 +1,278 @@
+{
+    "bomFormat": "CycloneDX",
+    "specVersion": "1.4",
+    "serialNumber": "urn:uuid:43840228-3fae-4d88-95aa-96512d2510dd",
+    "version": 1,
+    "metadata": {
+        "timestamp": "2023-05-03T18:19:41Z",
+        "tools": [
+            {
+                "vendor": "anchore",
+                "name": "syft",
+                "version": "0.47.0"
+            }
+        ],
+        "component": {
+            "bom-ref": "6b8edfe5f2756e0",
+            "type": "file",
+            "name": "/var/lib/containers/storage/vfs/dir/517aef0ffe20db360d19aa475dbbfbe03f452f53403881a31f9a475c83af788b"
+        }
+    },
```
Observation: we prefer the Syft SBOM for everything except `components` (for those, we merge but prefer cachi2)

The merging is meant to be a temporary solution, right? Otherwise there's some more to think about IMO, like mismatched specVersion (note that 1.5 is already released)

brunoapimentel replied:
Yeah, I think adding Cachi2 components to Syft's SBOM makes more sense, since Syft is the default one now (that exists in all pipelines regardless of them being hermetic).

Syft is still using spec 1.4 even in the lastest release, though  =/

chmeliik replied:
Using 1.4 is fine, the problem would be if we bump cachi2's specVersion, use some new features but keep 1.4 in the merged SBOM

chmeliik replied:
Taking the higher of the two specVersions would probably be fine, but again, I don't really need it in a temporary solution

brunoapimentel replied:
Hopefully we will have a better solution for multiple SBOMs by then.

chmeliik commented on tests/unit/data/sboms/cachi2.bom.json:
```diff
@@ -0,0 +1,99 @@
+{
+    "bomFormat": "CycloneDX",
+    "components": [
```
Would be good to add more metadata to the cachi2 test SBOM

ben-alkov commented:
More of an "I don't really have anything to add so I'll just talk" than a real comment :D

I have a deep fondness for certain Functional Programming principles. One can be summarized as "Single Entry, Single Exit"[1], f.e. this

```python
    if "purl" not in component:
        return component.get("name", "") + "@" + component.get("version", "")

    if "@" in component["purl"]:
        parts = component["purl"].split("@")
        return parts[0].lower() + "@" + quote_plus(parts[1])
    else:
        return component["purl"]
```

can be refactored into (roughly; there are some loose ends)

```python
    val = None

    if component.get("purl", None):
        if "@" in component["purl"]:
            parts = component["purl"].split("@")
            val = parts[0].lower() + "@" + quote_plus(parts[1])
        else:
            val = component["purl"]
    else:
        # No purl
        val = component.get("name", "") + "@" + component.get("version", "")

    return val
```

I feel like it usually ends up being a case of more logic vs. more `return`s

[1] Obviously, most languages enforce single *entry* by design, except under carefully controlled circumstances

ben-alkov commented:
@brunoapimentel;

LGTM so far - will NPM be a part of this PR?

brunoapimentel commented:
> @brunoapimentel;
> 
> LGTM so far - will NPM be a part of this PR?

Yes, I'll push some small changes later on to cover NPM.

brunoapimentel commented:
> More of an "I don't really have anything to add so I'll just talk" than a real comment :D
> 
> I have a deep fondness for certain Functional Programming principles. One can be summarized as "Single Entry, Single Exit"[1], f.e. this
> 
> ```python
>     if "purl" not in component:
>         return component.get("name", "") + "@" + component.get("version", "")
> 
>     if "@" in component["purl"]:
>         parts = component["purl"].split("@")
>         return parts[0].lower() + "@" + quote_plus(parts[1])
>     else:
>         return component["purl"]
> ```
> 
> can be refactored into (roughly; there are some loose ends)
> 
> ```python
>     val = None
> 
>     if component.get("purl", None):
>         if "@" in component["purl"]:
>             parts = component["purl"].split("@")
>             val = parts[0].lower() + "@" + quote_plus(parts[1])
>         else:
>             val = component["purl"]
>     else:
>         # No purl
>         val = component.get("name", "") + "@" + component.get("version", "")
> 
>     return val
> ```

I personally tend to use returns more to avoid to have control structures, specially if there needs to be nesting.

> I feel like it usually ends up being a case of more logic vs. more `return`s

Yeah, I agree. I guess it is probably better when the principle is language enforced.

I am trying to use more functional principles, btw, specially when I'm writing Java :smile:

chmeliik commented:
> I am trying to use more functional principles, btw, specially when I'm writing Java :smile:

Obligatory endorsement for Haskell  :heart: 

I couldn't resist and had to publish my Brainfuck interpreter that I made last year :grin: https://github.com/chmeliik/Brainfuck

brunoapimentel commented:
> > I am trying to use more functional principles, btw, specially when I'm writing Java smile
> 
> Obligatory endorsement for Haskell heart
> 
> I couldn't resist and had to publish my Brainfuck interpreter that I made last year grin https://github.com/chmeliik/Brainfuck

#### Why?
Why not?

brunoapimentel commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,164 @@
+    return component.get("purl", "").startswith("pkg:npm") and (
+        version.startswith("https://")
+        or version.startswith("git+https://")
+        or version.startswith("github:")
```
I guess that `bitbucket:` and `gitlab:` might also be valid in purls, but I still couldn't get a Syft SBOM with these type of dependencies in it.

chmeliik replied:
Maybe we just check if the version is any url by checking if there's a `:`? (Same logic as https://github.com/containerbuildsystem/cachi2/pull/229#discussion_r1259464017)

brunoapimentel replied:
> Maybe we just check if the version is any url by checking if there's a :
Yeah, that might do it :)

brunoapimentel replied:
I'll double check the behavior for v3 lockfiles.

brunoapimentel replied:
Syft does use only the version for lockfiles v2/v3 :facepalm: 

So this means we need to either:
- fallback to the name-only deduping method we are using for Pip
- live with the duplicates
- update Cachi2 so it also outputs the version for file/vcs dependencies

lkolacek commented on utils/merge_syft_sbom.py:
```diff
@@ -0,0 +1,154 @@
+            or key in cachi2_indexed_components.keys()
+        )
+
+    return component_is_duplicated
```
How the component is passed to function `component_is_duplicated(component: dict[str, Any])` here?

brunoapimentel replied:
It is actually used here: 
https://github.com/containerbuildsystem/cachi2/blob/ea97c3dbada30ef0527bf4664a72350c749e55a9/utils/merge_syft_sbom.py#L129-L133

The catch here is that the `component_is_duplicated` function that gets generated holds the info on the existing Cachi2 non-registry [components](https://github.com/containerbuildsystem/cachi2/blob/ea97c3dbada30ef0527bf4664a72350c749e55a9/utils/merge_syft_sbom.py#L96).

lkolacek replied:
Aha thanks :)

lkolacek reviewed (approved):
As far as I can tell looks great! :+1: Left just 1 minor to educate myself.

ben-alkov reviewed (commented):
LGTM, but I'd rather Adam have a last look at it before final approval and merge

brunoapimentel commented:
Just FYI: there's apparently a Syft bug that makes it report Pip dependencies twice, adding `%20\\` to the second one ([e.g.](https://github.com/brunoapimentel/rhtap-cachi2-sbom-merge/blob/main/pip/cachi2/rhtap.bom.json#L2074)).

I ignored it for now, since it seems to be fixed in the newer Syft versions.

taylormadore reviewed (approved):
As far as I can tell (I'm not too familiar with Syft), this LGTM
